README.md
# Quick Start

* Add PrometheusIntegration to the mods/ directory of your server.
  You can download it from https://madoka.brage.info/release/

* Install docker on your server, by whichever means.

  Debian/Ubuntu: sudo apt-get install docker.io

* Run the following commands.

  This will expose a dashboard on port 3000. Be careful, by default
  this is publicly writable; you'll need a reverse proxy such as nginx
  to lock it down. See the next section for hints as to how.

  The Prometheus instance on port 9090 also needs to be publicly
  accessible. Do not firewall it.

<!-- -->

    MCDASH_PORT=3000
    MCPROM_PORT=9090
    docker run -d --name mcprom -p $MCPROM_PORT:9090 baughn/mcprom
    docker run -d --name mcdash --link mcprom:mcprom -p $MCDASH_PORT:3000 baughn/mcdash

* Finally, visit http://(your hostname):3000/servers and edit the URL
  of the preconfigured Minecraft server to match the public hostname
  of your server.

# Optional follow-up

* Do the same thing client-side. It isn't as useful, but
  Promdash-Integration absolutely works in single-player or for
  multiplayer clients; the same instructions apply.

* Play around with Prometheus. The easiest way is to visit the /graph
  URL of your Prometheus instance; you can do the same with Promdash,
  but it's a little fiddlier.

  You can find more documentation at
  http://prometheus.io/docs/introduction/overview/

* Suggest improvements. Talk to me!

* Set up Nginx as a reverse proxy for Promdash, so other people can't
  edit your settings. While it shouldn't be possible to break out of
  the container, or even into the container, they can absolutely mess
  up your console.

  Here's a good stanza to start with, though note that your Prometheus
  port still needs to be accessible. Be careful, HTTP basic
  authentication sends your password in cleartext, so it is not safe
  to re-use passwords from elsewhere; ideally, use SSL instead.

<!-- -->

    server {
        listen 80;
        server_name promdash.example.com;

        location / {
            proxy_pass http://localhost:3000;
            limit_except GET HEAD {
                auth_basic "Promdash";
                auth_basic_user_file conf.d/htpasswd;
            }
        }
    }


changelog.md
## 0.2.2

Removed Jetty in favor of the built-in HttpServer.

## 0.2.1

Add quantiles to server/world tick times: 50th, 90th and 95th percentile.

## 0.2.0

First possibly-working 1.10.2 version


src/main/java/info/brage/minecraft/prometheus/Events.java
package info.brage.minecraft.prometheus;

import cpw.mods.fml.common.eventhandler.SubscribeEvent;
import cpw.mods.fml.common.gameevent.TickEvent;
import io.prometheus.client.Counter;
import io.prometheus.client.Gauge;
import io.prometheus.client.Summary;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.server.MinecraftServer;
import net.minecraft.world.World;
import net.minecraftforge.event.terraingen.ChunkProviderEvent;

import java.util.HashSet;

/**
 * Created by svein on 25/04/15.
 */
public class Events {
    static final Counter ticks = Counter.build()
            .name("ticks").help("Server ticks").register();
    static final Summary tickTime = Summary.build()
            .name("tick_time").help("Server tick time").register();
    static final Summary worldTickTime = Summary.build()
            .name("world_time").labelNames("world").help("World tick time").register();
    static final Counter worldgen = Counter.build()
            .name("worldgen").help("Chunks generated").register();
    static final Gauge present = Gauge.build()
            .name("present").labelNames("name").help("Is player present").register();

    Summary.Timer serverTimer;
    Summary.Timer worldTimer;

    int serverTicks = 0;

    @SubscribeEvent(receiveCanceled = true)
    public void onServerTick(TickEvent.ServerTickEvent event) {
        serverTicks++;

        if (event.phase == TickEvent.Phase.START) {
            serverTimer = tickTime.startTimer();
        } else if (event.phase == TickEvent.Phase.END) {
            serverTimer.observeDuration();
            serverTimer = null;
            ticks.inc();
        }

        if (serverTicks % 60 == 0) {
            setPlayers(MinecraftServer.getServer().getEntityWorld());
        }
    }

    @SubscribeEvent(receiveCanceled = true)
    public void onWorldTick(TickEvent.WorldTickEvent event) {
        if (event.phase == TickEvent.Phase.START) {
            worldTimer = worldTickTime.labels(event.world.provider.getDimensionName()).startTimer();
        } else if (event.phase == TickEvent.Phase.END) {
            worldTimer.observeDuration();
            worldTimer = null;
        }
    }

    @SubscribeEvent(receiveCanceled = true)
    public void onWorldGen(ChunkProviderEvent.ReplaceBiomeBlocks event) {
        worldgen.inc();
    }

    HashSet<String> seenPreviously = new HashSet<String>();

    void setPlayers(World world) {
        HashSet<String> seenNow = new HashSet<String>();

        // Set all new players' values to 1.
        for (Object obj : world.playerEntities) {
            if (obj instanceof EntityPlayer) {
                EntityPlayer player = (EntityPlayer)obj;
                String name = player.getDisplayName();
                seenNow.add(name);
                if (!seenPreviously.contains(name)) {
                    present.labels(name).set(1);
                }
            }
        }
        // Delete all gone-away players.
        for (String name : seenPreviously) {
            if (!seenNow.contains(name)) {
                present.remove(name);
            }
        }
        seenPreviously = seenNow;
    }
}


src/main/java/info/brage/minecraft/prometheus/PrometheusIntegration.java
package info.brage.minecraft.prometheus;

import cpw.mods.fml.common.FMLCommonHandler;
import cpw.mods.fml.common.Mod;
import cpw.mods.fml.common.event.FMLInitializationEvent;
import cpw.mods.fml.common.event.FMLPreInitializationEvent;
import io.prometheus.client.exporter.MetricsServlet;
import io.prometheus.client.hotspot.DefaultExports;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.common.config.Configuration;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.servlet.ServletContextHandler;
import org.eclipse.jetty.servlet.ServletHolder;

@Mod(modid = PrometheusIntegration.MODID, version = PrometheusIntegration.VERSION, acceptedMinecraftVersions="[1.7.2,)", acceptableRemoteVersions = "*")
public class PrometheusIntegration
{
    public static final String MODID = "prometheus-integration";
    public static final String VERSION = "1.0";

    int jettyPort;

    @Mod.EventHandler
    public void preInit(FMLPreInitializationEvent event) {
        Configuration config = new Configuration(event.getSuggestedConfigurationFile());
        config.load();
        jettyPort = config.getInt("jetty", "ports", 1234, 1025, 32767, "Port to run the metrics-server on.");
        config.save();
    }
    
    @Mod.EventHandler
    public void init(FMLInitializationEvent event) throws Exception
    {
        // Get port config.


        // Start up Jetty.
        Server server = new Server(jettyPort);
        ServletContextHandler context = new ServletContextHandler();
        context.setContextPath("/");
        server.setHandler(context);
        context.addServlet(new ServletHolder(new MetricsServlet()), "/metrics");
        server.start();

        // Get some Hotspot stats.
        DefaultExports.initialize();

        // Connect to FML.
        MinecraftForge.EVENT_BUS.register(new Events());
        Events events = new Events();
        FMLCommonHandler.instance().bus().register(events);
        //MinecraftForge.TERRAIN_GEN_BUS.register(events);
    }
}



src/main/kotlin/info/brage/minecraft/prometheus/PrometheusIntegration.kt
package info.brage.minecraft.prometheus

import io.prometheus.client.Counter
import io.prometheus.client.Gauge
import io.prometheus.client.Summary
import io.prometheus.client.exporter.HTTPServer
import io.prometheus.client.hotspot.DefaultExports
import net.minecraft.entity.player.EntityPlayer
import net.minecraftforge.common.ForgeChunkManager
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.common.config.Configuration
import net.minecraftforge.event.terraingen.PopulateChunkEvent
import net.minecraftforge.fml.common.FMLCommonHandler
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.common.event.FMLInitializationEvent
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.fml.common.gameevent.TickEvent
import java.util.*


const val MODID = "prometheus-integration"
const val VERSION = "1.0"

@Mod(modid = MODID, version = VERSION, acceptedMinecraftVersions="[1.10.2,)", acceptableRemoteVersions = "*")
class PrometheusIntegration {

    private var httpPort: Int = 0

    @Mod.EventHandler
    fun preInit(event: FMLPreInitializationEvent) {
        val config = Configuration(event.suggestedConfigurationFile)
        config.load()
        httpPort = config.getInt("jetty", "ports", 1234, 1025, 32767, "Port to run the metrics-server on.")
        config.save()
    }

    @Mod.EventHandler
    fun init(event: FMLInitializationEvent) {
        val server = HTTPServer(httpPort, true)

        // Get some Hotspot stats.
        DefaultExports.initialize()

        // Connect to Forge.
        MinecraftForge.EVENT_BUS.register(eventHandler)
        MinecraftForge.TERRAIN_GEN_BUS.register(eventHandler)
    }
}


object eventHandler {
    // Tick rate.
    val ticks = Counter.build().name("ticks").help("Server ticks").register()
    val tickTime = Summary.build().name("tick_time").help("Server tick time")
        .quantile(0.5, 0.05)
        .quantile(0.9, 0.01)
        .quantile(0.95, 0.01)
        .register()
    val worldTickTime = Summary.build().name("world_time").labelNames("world").help("World tick time")
        .quantile(0.5, 0.05)
        .quantile(0.9, 0.01)
        .quantile(0.95, 0.01)
        .register()
    // World generation.
    val worldgen = Counter.build().name("worldgen").help("Chunks generated").register()
    // Players.
    val present = Gauge.build().name("present").labelNames("name").help("Is player present").register()
    // Possible load causes.
    val chunksLoaded = Gauge.build().name("chunks_loaded").labelNames("cause", "world")
            .help("Count of chunks loaded, by cause").register()

    val server by lazy {
        FMLCommonHandler.instance().minecraftServerInstance
    }

    var serverTimer: Summary.Timer? = null
    var worldTimer: Summary.Timer? = null

    var serverTicks = 0

    private fun setLoadedChunks() {
        for (world in server.worlds) {
            val total = world.chunkProvider.loadedChunkCount
            val forced = ForgeChunkManager.getPersistentChunksFor(world).size()
            val name = world.provider.dimensionType.getName()
            chunksLoaded.labels("chunkloader", name).set(forced.toDouble())
            chunksLoaded.labels("other", name).set((total - forced).toDouble())
        }
    }

    @SubscribeEvent
    fun onServerTick(event: TickEvent.ServerTickEvent) {
        serverTicks++

        if (event.phase == TickEvent.Phase.START) {
            serverTimer = tickTime.startTimer()
        } else if (event.phase == TickEvent.Phase.END) {
            serverTimer!!.observeDuration()
            serverTimer = null
            ticks.inc()
        }

        when (serverTicks % 60) {
            27 -> setLoadedChunks()
            47 -> setPlayers()
        }
    }

    @SubscribeEvent
    fun onWorldTick(event: TickEvent.WorldTickEvent) {
        if (event.phase == TickEvent.Phase.START) {
            worldTimer = worldTickTime.labels(event.world.provider.dimensionType.getName()).startTimer()
        } else if (event.phase == TickEvent.Phase.END) {
            worldTimer!!.observeDuration()
            worldTimer = null
        }
    }

    @SubscribeEvent
    fun onWorldGen(event: PopulateChunkEvent.Pre) {
        worldgen.inc()
    }

    var seenPreviously = HashSet<String>()

    fun setPlayers() {
        val seenNow = HashSet<String>()

        // Set all new players' values to 1.
        for (world in server.worlds) {
            for (obj in world.playerEntities) {
                if (obj is EntityPlayer) {
                    val name = obj.displayNameString
                    seenNow.add(name)
                    if (!seenPreviously.contains(name)) {
                        present.labels(name).set(1.0)
                    }
                }
            }
        }
        // Delete all gone-away players.
        for (name in seenPreviously) {
            if (!seenNow.contains(name)) {
                present.remove(name)
            }
        }
        seenPreviously = seenNow
    }

}
